### DISCUSSION TITLE
How to get data from my RealTimeDatabase ?

### DISCUSSION THREAD

**Discussion by TheArtyMorty** [GodotNuts/GodotFirebase#429] [Q&A] [answered]
Hello,
This is my first time using Godot and GodotFirebase, so I may be missing something obvious. Sorry for the inconvenience if that's the case. 

Important note : I'm using Godot 4, so I got the code from the 4.x branch which seems to works fine. Don't know if it could be an issue for me.
I'm trying to query data from my RealTimeDatabase, but I can't ever get the signal new_data_update to trigger. 
Here is my code : 

```
extends Node
	
# Variables
var db_ref

func _ready():
	Firebase.Auth.connect("login_succeeded", Callable(self,  "_on_FirebaseAuth_login_succeeded"))
	Firebase.Auth.connect("login_failed",  Callable(self, "on_login_failed"))
	Firebase.Auth.login_with_email_and_password("test@gmail.com", "123456")
	#Firebase.Auth.login_anonymous() # FOR SOME REASON, ANONYMOUS LOGIN DOESN'T TRIGGER LOGGIN_SUCCEEDED SIGNAL

func _on_FirebaseAuth_login_succeeded(_auth):
	print("login succeeded, we try to query the database")
	db_ref = Firebase.Database.get_database_reference("Tours", {})
	db_ref.connect("new_data_update", Callable(self, "_on_db_data_update"))
	db_ref.push({"username": "toto"})   #THIS WORKS FINE, I SEE IT POP IN MY DATABASE

func on_login_failed(error_code, message):
	print("error code: " + str(error_code))
	print("message: " + str(message))

func _on_db_data_update(resource):
	#I NEVER GET HERE... WHAT AM I MISSING?
	print("We got something !!!!")
	print(resource)

```

I followed the https://github.com/GodotNuts/GodotFirebase/wiki/Realtime-Database#connecting-signals examples, and after some struggles I was able to make it partially works : 
The login succeed, the db_ref is valid and I'm able to delete or push data to it. 
**However, the new_data_update signal is never called. What did I miss?** 

Bonus question : I first tried to login anonymously, and even after I activated that on my firebase project, it never seemed to reach _on_FirebaseAuth_login_succeeded. I wasn't getting on_login_failed anymore, but no succeed either. Logging with an email does work fine though.

**Comment by WolfgangSenff** [comment]
I'm not sure why anonymous login wouldn't work. But there's an obvious problem why the realtime database isn't working: you are calling to the reference immediately. It needs to wait a little bit to make the connection the first time. There's also a much better way to connect to the signal. Here, try this:

```
func _on_FirebaseAuth_login_succeeded(_auth):
  db_ref = Firebase.Database.get_database_reference("Tours", {})
  db_ref.new_data_update.connect(_on_db_data_update)
  await get_tree().create_timer(0.1).timeout
  db_ref.push({"username": "toto"})
```

Everything else can stay the same I believe. If that doesn't work, my only real guess based on the info you've provided is something must be wrong with the configuration, probably the rules. Regarding anonymous login, I suspect that's the case as well. Our tester program runs all of these just fine. https://github.com/GodotNuts/GodotFirebaseTestHarness/

**Comment by TheArtyMorty** [comment]
Hello @WolfgangSenff, thanks for the quick feedback. 
Unfortunately, that did not help, although it does seems to be a much easier way to connect to the signal. 

I tried the code you sent and got the exact same result : 
- my "_on_db_data_update" function wasn't called. 
- the push properly occured and the data appeared on my realtime database. 

My rules for the database couldn't be simplier : 
```
{
  /* Visit https://firebase.google.com/docs/database/security to learn more about security rules. */
  "rules": {
    ".read": true,
    ".write": true
  }
}
```

So the db_ref looks fine, the login looks fine (at least the succeded signal was sent and caught), the rules looks fine. 
Why don't I get a signal for new_data_update ?

**Comment by WolfgangSenff** [comment]
Hm, I'm not sure. I'd probably have to see more of your setup. Is it possible to PM me the project (without the .env file) on Discord? I saw you just joined. I'm happy to take a look that way.

**Comment by TheArtyMorty** [comment, marked as answer]
@WolfgangSenff , nevermind I'm stupid. I was playing with the addons files to understand and debug, like returning error code for some requests and so on and I messed up something. Once I reverted my local changes, it works fine properly. 

**The solution for my case, which I'll mark as solution, was to not mess up the addons file, they work perfectly as is 👍** 

Side note : Once that was working, I tested again the anonymous login, and that still doesn't work. I'm not hitting the _on_FirebaseAuth_login_succeeded. 
On firebase the anonymous user is properly created though.
Maybe that's because I should listen to the register signal instead?
